### PR TITLE
settings-specs

### PR DESCRIPTION
**Pull request by maddiemlucas**
# Data Plane API - Pull Request
Added swagger documentation to support the new Settings API which allows users to update, get or list pool settings. 

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

https://microsoft-my.sharepoint.com/:w:/p/maddielucas/EXRbV7zlAlVDoQNNt9rMrrEBtRn2DDRd7PX2gv9jpEFGuQ?e=qBLruE

Is this review for (select one):

- a public preview

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document: https://microsoft-my.sharepoint.com/:w:/p/maddielucas/EXRbV7zlAlVDoQNNt9rMrrEBtRn2DDRd7PX2gv9jpEFGuQ?e=qBLruE

* Updated paths:
* /settings
* /settings/{setting-name}

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            },
+            "default": {
+              "description": "Key Vault error response describing why the operation failed.",
+              "schema": {
+                "$ref": "common.json#/definitions/KeyVaultError"
+              }
+            }
+          },
+          "x-ms-examples": {
+            "Create key": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:76]: "Update setting"?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            "Settings"
+          ],
+          "operationId": "GetSettingValue",
+          "summary": "Get specified account setting value.",
+          "description": "Retrieves the value of a specified, value account setting.",
+          "parameters": [
+            {
+              "$ref": "#/parameters/ApiVersionParameter"
+            }
+          ],
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:92]: As named and described, this gets a single setting value but takes no setting name. You should define the "setting-name" parameter in the "parameters" section of this file (add it) and `$ref` it in both places.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            "default": {
+              "description": "Key Vault error response describing why the operation failed.",
+              "schema": {
+                "$ref": "common.json#/definitions/KeyVaultError"
+              }
+            }
+          },
+          "x-ms-pageable": {
+            "nextLinkName": "nextLink"
+          },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:109]: How is getting a single setting pageable?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+          "x-ms-examples": {
+            "GetKeys": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:111]: "Get setting"?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:148]: "Get settings"?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+          }
+        }
+      }
+    },
+    "definitions": {
+      "UpdateSettingsRequest": {
+        "properties": {
+          "value": {
+            "type": "string",
+            "description": "The account setting value. True if setting is enabled, false if disabled"
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:160]: It's a string named "value", so why would someone pass boolean values? Seems there should be a `enabled` property that should be true or false, yeah?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+              "type": "string",
+              "description": "The name of the account setting. Must be a valid settings option."
+            },
+            {
+              "name": "parameters",
+              "in": "body",
+              "required": true,
+              "x-ms-client-flatten": true,
+              "schema": {
+                "$ref": "#/definitions/UpdateSettingsRequest"
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:53]: After reviewing models, I don't understand how this is used. It's called "UpdateSetting" but apparently you don't update a setting - just enable or disable it. So at least now I see what your description refers to but I'm now confused on how a setting value is even changed. Is that strictly a management plane operation? If so, why is data plane enabling or disabling values? It's a very weird mix.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+      },
+      "Setting": {
+        "properties": {
+          "name": {
+            "type": "string",
+            "description": "The account setting to be updated"
+          },
+          "value": {
+            "type": "string",
+            "description": "The account setting value. True if setting is enabled, false if disabled"
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:176]: Why is this defined as a string if you're describing boolean values?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+          }
+        },
+        "required": [
+          "name",
+          "value"
+        ],
+        "description": "The update settings response object containing the setting name and its value."
+      },
+      "SettingsListResult": {
+        "properties": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:186]: This needs to define a `nextLink` property as well, per your pageable definition. Generally, all list operations should be pageable unless a team is 100% sure that a returned array will never, ever be long enough to be paged. That's not typical. See https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#collections

**Review by heaths — CHANGES_REQUESTED** [review]
There's a number of things that need to be fixed, and you should schedule a REST API Stewardship board review at some point. We require it for GAs, but previews are encouraged, especially when there's a lot of questions/confusions about the feature like I have here. See https://aka.ms/azapi for more information.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json (hunk 8 above)
Comment (maddiemlucas) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:176]: Right now, we only have one setting that only needs to be Off/On or in other words True/False. In the future, we will likely need settings that can hold string values and not be limited to true or false. I will update this description though to not limit it to only a "boolean" type definition.

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:176]: Changing the `type` to something like `object` is not going to make for a good API. Off hand, I would say this needs to be a polymorphic type but having derivatives with just a single typed value seems overkill. Please set up a meeting - or come to office hours (for something this small, that's probably better) - with the API Stewardship board.

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:176]: Alternatively, look at AppConfiguration: https://github.com/Azure/azure-rest-api-specs/blob/4b5fe2fb0a5066c4ff2bd429dbd5e1afda6afab3/specification/appconfiguration/data-plane/Microsoft.AppConfiguration/stable/1.0/appconfiguration.json#L1194-L1226

That said, this still isn't a great experience for client libraries since strong-typed languages will need to provide an abstraction over the type when we're trying to move to more generated code by default.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json (hunk 7 above)
Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:53]: So the value here is a boolean because the only settings suppored are boolean? But that also doesn't seem like enabling/disabling. If simple enabling/disabling a setting is what you're doing here, then use `enabled`. If the only supports values are boolean, 1) don't mention "enabling" or "disabling" the setting because that's not what this is doing, it's setting a value that happens to be a boolean, and 2) see my response below about this API supporting other types in the future. It has to remain compatible initially, so some sort of `type` property is necessary to describe the type like AppConfiguration did. At the very least, client libraries can parse/emit any intrinsic type as a string, but this also may not be ideal for code generation.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 10]
```diff
@@ -86,6 +86,13 @@
           "summary": "Get specified account setting value.",
           "description": "Retrieves the value of a specified, value account setting.",
           "parameters": [
+            {
+              "name": "setting-name",
+              "in": "path",
+              "required": true,
+              "type": "string",
+              "description": "The name of the account setting. Must be a valid settings option."
+            },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:95]: Can you add this to "parameters" and just ref it in both places, like you ref the ApiVersionParameter from common.json (just put the new one in this doc, though). That's the best way to make sure the definition remains consistent across endpoints.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json (hunk 9 above)
Comment (maddiemlucas) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:183]: We are 100% it will never be pageable.

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:183]: Then you need to get rid of the x-pageable on line 140.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json (hunk 8 above)
Comment (maddiemlucas) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json]: Office hours discussion: Agreed on adding a type specifier to support multiple intrinsic types 

```
"value": {
  "type": "string"
},
"type": {
  "enum": [
    "boolean",
    "string",
    "integer",
    "float"
  ]
}
```

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 11]
```diff
@@ -171,6 +168,16 @@
           "value": {
             "type": "string",
             "description": "The value of the pool setting."
+          },
+          "type": {
+            "type": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:173]: Much better, but:

1. Mark this as `modelAsString: true`: https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-enum
2. The list I gave like this in the meeting was just an example. If you only want to support boolean for now, you can as long as you mark it as `modelAsString: true`. The values can be updated then as you need to, right @JeffreyRichter?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 12]
```diff
@@ -137,9 +137,6 @@
               }
             }
           },
-          "x-ms-pageable": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json]: During a different sync meeting later this afternoon, I think it was @vickm who suggested that maybe this should be pageable initially, which means the response needs a `nextLink`. Effectively, how the generated client SDK methods look will be very different - and breaking if changed - if this is pageable or not. So if there's even a remote chance of paging in the future, it needs to be pageable now.

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json (hunk 12 above)
Comment (maddiemlucas) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json]: Ok, we all just synced up and decided that it will never be pageable. The max number of items that it could ever even get to right now is 6 and we still don't even foresee that happening.

**Review by jlichwa — APPROVED** [review]

**Review by heaths — APPROVED** [review]

**Comment by heaths** [comment]
@maddiemlucas can you close and re-open this PR to see if you can get the pipelines unstuck? This sometimes helps.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 13]
```diff
@@ -170,19 +169,13 @@
             "description": "The value of the pool setting."
           },
           "type": {
-            "type": {
               "enum": [
                 "boolean"
               ],
             "modelAsString": true,
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:175]: This has to go into an `x-ms-enum`. See https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-enum

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 14]
```diff
@@ -170,19 +169,13 @@
...
             "description": "The type specifier of the value."
           }
-        },
-        "required": [
-          "name",
-          "value"
-        ],
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json]: Why did you remove this? A name and value certainly seem like required properties to set the value of a named setting. This information is necessary not only to set expectations for REST API callers, but for the generated SDK clients as well.

**Review by heaths — CHANGES_REQUESTED** [review]

**Review by heaths — APPROVED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json [hunk 15]
```diff
@@ -5,6 +5,7 @@
     "description": "The key vault client performs cryptographic key operations and vault operations against the Key Vault service.",
     "version": "7.4-preview.1"
   },
+  "x-ms-parameter-location": "client",
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/common.json:8]: This needs to go on the ApiVersionParameter itself, not just arbitrarily in the swagger. See https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-parameter-location

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+  },
+  "parameters": {
+    "ApiVersionParameter": {
+      "name": "api-version",
+      "in": "query",
+      "required": true,
+      "type": "string",
+      "description": "Client API version.",
+      "x-ms-parameter-location": "client"
+    },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1/settings.json:212]: Use the one in common.json once you've added the `x-ms-parameter-location` appropriately.

**Review by heaths — CHANGES_REQUESTED** [review]

**Review by heaths — APPROVED** [review]

**Comment by heaths** [comment]
@lmazuel @weshaggard can you override and merge? The ModelValidation errors we've seen before about double slashes, which in practice don't cause an issue. Also getting erroneous errors saying "missing required property" when, in fact, they are in the sample correctly. Not sure what's going on there.
